dFusion/dFusionSpec.md

```diff
@@ -6,34 +6,69 @@ A specification developed by Gnosis.
+
+The matching process is decentralized.
+After orders have been collected over a certain amount of time, a batch is frozen and any sufficiently bonded participant can suggest a matching of orders in the current batch.
+
+If more than one solution is submitted within a certain amount of time after the batch closes, the one that generates the largest "trader surplus" (detailed explanation below, for now think "trading volume") is selected and executed.
+For this, the "winner" has to post enough information about the solution on-chain, so that any participant can quickly check the validity of the solution.
+
+Anyone can challenge a solution on-chain (while providing a bond and an alternative solution) within a certain time-period after posting.
+Such disputes are resolved by verifying a zkSnark proof on-chain that checks if the provided solution fulfills a number of constraints (specified in detail below).
+If the verification succeeds, the challenger loses their bond and the state is "finalized". 
+If the verification fails the provided alternative state is assumed correct unless challenged within a certain time.
+
+## State stored in the smart contract
+
+For each account, we chain each of ERC20 token balance together and store them as pedersen hash (not merkleized) in the anchor smart contract.
+This "compressed" representation of all account balances is collision resistant and can thus be used  which will store all relevant information for this snapp exchange.
 The following diagram shows the state construction:
 
 ![State construction](./dFusion%20rootHash.png?raw=true "State construction")
 
-To allow `K` to be small, a bi-map of an accounts public key (on-chain address) to its `accountIndex` will be stored in the anchor contract as well. Accounts will pay "rent" to occupy an active account. The account index can be used to locate a users token balances in the state.
+To allow `K` to be small, a bi-map of an accounts public key (on-chain address) to its `accountIndex` will be stored in the anchor contract as well. 
+Accounts will pay "rent" to occupy an active account. The account index can be used to locate a users token balances in the state.
 
-All orders are encoded as limit sell orders: `(accountIndex, fromTokenIndex, toTokenIndex, limitPrice, amount, batchId, signature)`.
-The order should be read in the following way: the user occupying the specified *accountIndex* would like to sell the token *fromTokenIndex* for *toTokenIndex* for at most the *limitPrice* and the *amount* specified.
-The *batchId* and *signature* allow a third party to submit an order on behalf of others (saving gas when batching multiple orders together).
-The user only has to specify which batch their order is valid for and sign all the information with their private key.
+Furthermore, we store a bi-map of token address to its index `0 <= n <= N`, for each token that can be traded on the exchange.
```

Comment:
This is picky, but we may want to consider using `T` for number of tokens and reserve `N` for something else?